### PR TITLE
Simplify balance further

### PR DESCRIPTION
**Pull request by vkuznecovas**
After discussions, decided to add a struct wrapping the calculations,
as well as emitting a single grand total change event for simplicity.

### CODE REVIEW

File: session/pingpong/consumer_balance_tracker.go [hunk 1]
```diff
@@ -129,8 +125,13 @@ func (cbt *ConsumerBalanceTracker) handleUnlockEvent(id string) {
 	cbt.ForceBalanceUpdate(identity)
 }
 
-func (cbt *ConsumerBalanceTracker) handleGrandTotalRecovered(ev GrandTotalRecovered) {
-	cbt.ForceBalanceUpdate(ev.Identity)
+func (cbt *ConsumerBalanceTracker) handleGrandTotalChanged(ev AppEventGrandTotalChanged) {
+	if _, ok := cbt.balances[ev.ConsumerID]; !ok {
```
Comment (anjmao) [inline session/pingpong/consumer_balance_tracker.go]: Missing lock?

File: session/pingpong/consumer_balance_tracker.go [hunk 2]
```diff
@@ -129,8 +125,13 @@ func (cbt *ConsumerBalanceTracker) handleUnlockEvent(id string) {
 	cbt.ForceBalanceUpdate(identity)
 }
 
-func (cbt *ConsumerBalanceTracker) handleGrandTotalRecovered(ev GrandTotalRecovered) {
-	cbt.ForceBalanceUpdate(ev.Identity)
+func (cbt *ConsumerBalanceTracker) handleGrandTotalChanged(ev AppEventGrandTotalChanged) {
```
Comment (Waldz) [inline session/pingpong/consumer_balance_tracker.go]: I suggest to pass to what amount it changed in `AppEventGrandTotalChanged `, and you would need to call `consumerGrandTotalGetter.Get()` or couple to this storage at all

File: session/pingpong/consumer_balance_tracker.go [hunk 3]
```diff
@@ -214,34 +216,36 @@ func (cbt *ConsumerBalanceTracker) handleStopEvent() {
...
+		cbt.balances[id] = v
+	} else {
+		cbt.ForceBalanceUpdate(id)
 	}
 
-	current := safeAdd(before, b)
-	cbt.balances[id] = current
-
-	go cbt.publishChangeEvent(id, before, current)
+	go cbt.publishChangeEvent(id, before, cbt.balances[id].GetBalance())
```
Comment (Waldz) [inline session/pingpong/consumer_balance_tracker.go]: Maybe publish full struct as it changes

File: session/pingpong/invoice_payer_test.go [hunk 4]
```diff
@@ -689,13 +700,19 @@ func TestInvoicePayer_issueExchangeMessage(t *testing.T) {
...
 	err        error
 	calledWith uint64
 }
 
-func (mcts *mockConsumerTotalsStorage) Store(providerAddress, accountantAddress string, amount uint64) error {
+func (mcts *mockConsumerTotalsStorage) Store(consumerAddress, accountantAddress string, amount uint64) error {
 	mcts.calledWith = amount
+	go mcts.bus.Publish(AppTopicGrandTotalChanged, AppEventGrandTotalChanged{
+		Current:      amount,
+		AccountantID: identity.FromAddress(accountantAddress),
```
Comment (Waldz) [inline session/pingpong/invoice_payer_test.go]: `Identity` struct is construct of local MYST accounts, no point use this for `accountantAddress`

File: session/pingpong/consumer_totals_storage.go [hunk 5]
```diff
@@ -29,20 +31,34 @@ const consumerTotalStorageBucketName = "consumer_promised_totals"
...
-func NewConsumerTotalsStorage(bolt persistentStorage) *ConsumerTotalsStorage {
+func NewConsumerTotalsStorage(bolt persistentStorage, bus eventbus.Publisher) *ConsumerTotalsStorage {
 	return &ConsumerTotalsStorage{
 		bolt: bolt,
+		bus:  bus,
 	}
 }
 
 // Store stores the given amount as promised for the given channel.
 func (cts *ConsumerTotalsStorage) Store(consumerAddress, accountantAddress string, amount uint64) error {
```
Comment (Waldz) [inline session/pingpong/consumer_totals_storage.go]: Probably can pass `consumerID` your validated value in upper layer, now it recalculated second time

File: session/pingpong/invoice_payer_test.go (hunk 4 above)
Comment (vkuznecovas) [reply session/pingpong/invoice_payer_test.go]: There's a point, they format the addresses differently therefore all the addresses should be stored in identity format.

File: session/pingpong/consumer_totals_storage.go (hunk 5 above)
Comment (vkuznecovas) [reply session/pingpong/consumer_totals_storage.go]: what?

File: session/pingpong/consumer_balance_tracker.go (hunk 3 above)
Comment (vkuznecovas) [reply session/pingpong/consumer_balance_tracker.go]: I'll break a ton of users if i do that, so no, it's staying here :)

File: session/pingpong/consumer_balance_tracker.go (hunk 2 above)
Comment (vkuznecovas) [reply session/pingpong/consumer_balance_tracker.go]: I'm already passing in the amount.

File: session/pingpong/consumer_balance_tracker.go (hunk 1 above)
Comment (vkuznecovas) [reply session/pingpong/consumer_balance_tracker.go]: actually - no, both of the cases are internally locked aftewards, so we're fine here.

**Pull request merged**